### ISSUE TITLE
Could not use SFT Trainer in qlora_finetuning.py

### ISSUE THREAD

**Issue by shungyantham** [intel-analytics/ipex-llm#12356]
I have installed trl<0.12.0 to run qlora_finetune.py in the QLoRA/trl-example but it requires transformers 4.46.2 which causes the error below.
<img width="327" alt="incorrect transformer versio" src="https://github.com/user-attachments/assets/5d4ee4eb-5feb-4f3a-8ee7-eefa98f24947">

So I downgraded trl from 0.11.4 to 0.9.6 and I got another padding error.
<img width="605" alt="padding issue" src="https://github.com/user-attachments/assets/7d24f21a-4f38-4416-b92b-933a54908cfa">

**Comment by qiyuangong** [comment]
> I have installed trl<0.12.0 to run qlora_finetune.py in the QLoRA/trl-example but it requires transformers 4.46.2 which causes the error below. <img alt="incorrect transformer versio" width="327" src="https://private-user-images.githubusercontent.com/165883126/383823306-5d4ee4eb-5feb-4f3a-8ee7-eefa98f24947.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwMzI2MDgsIm5iZiI6MTczMTAzMjMwOCwicGF0aCI6Ii8xNjU4ODMxMjYvMzgzODIzMzA2LTVkNGVlNGViLTVmZWItNGYzYS04ZWU3LWVlZmE5OGYyNDk0Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwOFQwMjE4MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMDQyZjhmYTk5ZTY5NjExM2ZiZTJhOWQzOTBmYzUzZjI2OTNiY2ZkZDQwOTc0ZjA4Y2FhN2YyNmRiZDE5ODllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.oSxsZIlj0DLE-NgvII45I74Xinq6gaXTQRaSlRIHrEU">
> 
> So I downgraded trl from 0.11.4 to 0.9.6 and I got another padding error. <img alt="padding issue" width="605" src="https://private-user-images.githubusercontent.com/165883126/383826847-7d24f21a-4f38-4416-b92b-933a54908cfa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwMzI2MDgsIm5iZiI6MTczMTAzMjMwOCwicGF0aCI6Ii8xNjU4ODMxMjYvMzgzODI2ODQ3LTdkMjRmMjFhLTRmMzgtNDQxNi1iOTJiLTkzM2E1NDkwOGNmYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwOFQwMjE4MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wM2YwOTQ1OTVmYjIzZThlODVkZjdjMzg5MmNmY2RkMTgzZjBkMDdkYWVlODFkNzcyOGIwY2IyYjUxMjFhNjIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.EyAxs4118-WvLEjwt7yu0jFqosQ-GlUdM8f6NiZgcKY">

These errors is caused by transformers version mismatch. Can you downgrade transformers version to `4.36.0 `?

```bash
pip install transformers==4.36.0 datasets
```

**Comment by shungyantham** [comment]
Hi, I have also downgraded the transformers to 4.36.0 when I downgrade the trl to 0.9.6 and I got this error

**Comment by shungyantham** [comment]
https://github.com/intel-analytics/ipex-llm/blob/main/docker/llm/finetune/xpu/Dockerfile

I build this Dockerfile and then manually pip install trl==0.9.6 in the docker container. I ran the qlora_finetune.py in LLM_Finetuning/QLoRA/trl-example. Is there anything I missed?

**Comment by qiyuangong** [comment]
> https://github.com/intel-analytics/ipex-llm/blob/main/docker/llm/finetune/xpu/Dockerfile
> 
> I build this Dockerfile and then manually pip install trl==0.9.6 in the docker container. I ran the qlora_finetune.py in LLM_Finetuning/QLoRA/trl-example. Is there anything I missed?

Hi @shungyantham , we have reproduced this issue in our local env.

Please modify ` qlora_finetune.py` Line 91. Add ` data_collator=transformers.DataCollatorForSeq2Seq(
            tokenizer, pad_to_multiple_of=8, return_tensors="pt", padding=True
        )` to  SFTTrainer.

Code should look like this:

```python
    trainer = SFTTrainer(
        model=model,
        train_dataset=train_data,
        args=transformers.TrainingArguments(
            per_device_train_batch_size=4,
            gradient_accumulation_steps= 1,
            warmup_steps=20,
            max_steps=200,
            learning_rate=2e-5,
            save_steps=100,
            bf16=True,  # bf16 is more stable in training
            logging_steps=20,
            output_dir="outputs",
            optim="adamw_hf", # paged_adamw_8bit is not supported yet
            gradient_checkpointing=True, # can further reduce memory but slower
        ),
        dataset_text_field="instruction",
        data_collator=transformers.DataCollatorForSeq2Seq(
            tokenizer, pad_to_multiple_of=8, return_tensors="pt", padding=True
        )
    )
```

**Comment by qiyuangong** [comment]
https://github.com/intel-analytics/ipex-llm/pull/12368

**Comment by shungyantham** [comment]
Hi @qiyuangong , I have faced another issue after adding the padding to the Trainer
<img width="701" alt="image" src="https://github.com/user-attachments/assets/fd9b82a8-6868-4ca2-a286-ff506a95803a">

**Comment by qiyuangong** [comment]
> Hi @qiyuangong , I have faced another issue after adding the padding to the Trainer <img alt="image" width="701" src="https://private-user-images.githubusercontent.com/165883126/384755689-fd9b82a8-6868-4ca2-a286-ff506a95803a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMTI0ODEsIm5iZiI6MTczMTMxMjE4MSwicGF0aCI6Ii8xNjU4ODMxMjYvMzg0NzU1Njg5LWZkOWI4MmE4LTY4NjgtNGNhMi1hMjg2LWZmNTA2YTk1ODAzYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExMVQwODAzMDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00N2Y4OGI4OWZkYjRiNzgwOGMzYWQyN2JkM2YwMjdhZmFiNjYwZGQ2NzljMjk1NDU2MmE5MDUyMzBjMDQ3MzNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.35hXTBZA94rwQkjtdam9IQQKKzVf-ZIczAlPNDVDa4E">

Please provide transformers and trl version, as well as finetune.py.

**Comment by qiyuangong** [comment]
> Hi @qiyuangong ,
> 
> I setup a docker container following this link: https://github.com/intel-analytics/ipex-llm/tree/main/docker/llm/finetune/xpu
> 
> inside this container, I manually installed trl==0.9.6 and ran qlora_finetuning.py in /LLM-Finetuning/QLoRA/trl-example. My transformers version is 4.36.0.
> 
> <img alt="image" width="461" src="https://private-user-images.githubusercontent.com/165883126/385668738-e41dbacb-a212-46ac-8b17-a6612c4bc01d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0OTcwMTAsIm5iZiI6MTczMTQ5NjcxMCwicGF0aCI6Ii8xNjU4ODMxMjYvMzg1NjY4NzM4LWU0MWRiYWNiLWEyMTItNDZhYy04YjE3LWE2NjEyYzRiYzAxZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExM1QxMTE4MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMTU5NDRmNTQ1N2Y5N2RhMTA5YmQxODRiNzA4N2YxNjJhOTdlNDI0ZGUwYzQ0ZTIyYWFjZTBmYmE2ZWRlZTcxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.nFaxPpg0pJuTrO0DEpaPPBdMvLs8JfZJvvKVkCBdcmo">

Essential lib versions are correct. Please share your launch command and `qlora_finetuning.py`.

**Comment by qiyuangong** [comment]
We can resolve this padding error by adding these lines into `qlora_finetuning.py`.
```python
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token
```

I will submit a PR for this change.

**Comment by qiyuangong** [comment]
https://github.com/intel-analytics/ipex-llm/pull/12398
